File: desafio4/kotlin/src/main/kotlin/huffman/HuffTree.kt
package huffman

/**
 * Build HuffmanTree from frequency table of symbols
 * Created by ediaz on 5/21/17.
 */
abstract class HuffTree(val frequency : Int) {
}

class HuffLeaf(frequency: Int, val symbol: Char) : HuffTree(frequency) {

    fun symbolIndex(): Int = symbol.toInt() and 0xFFFF

}

val maxSymbols = 256

class HuffNode(val left: HuffTree, val right: HuffTree) : HuffTree(left.frequency+right.frequency)

class HuffHeap {

    var heap = arrayOfNulls<HuffTree>(maxSymbols+1)
    var last = 0

    fun insert(tree:HuffTree) {
        if (full()) {
            throw ArrayIndexOutOfBoundsException()
        }

        last++
        heap[last] = tree
        var j = last
        while (j > 1){
            if (heap[j]!!.frequency < heap[j/2]!!.frequency) {
                val tmp = heap[j]!!
                heap[j] = heap[j / 2]!!
                heap[j / 2] = tmp
            }
            j /= 2
        }
    }

    fun full() = last == maxSymbols

    fun empty() = last == 0

    fun size() = last

    fun extract() : HuffTree {
        if (empty()) {
            throw ArrayIndexOutOfBoundsException()
        }

        val min = heap[1]!!
        heap[1] = heap[last]
        last--
        var j = 1
        while (2*j <= last) {
            var k = 2*j
            if (k+1 <= last && heap[k+1]!!.frequency < heap[k]!!.frequency) {
                k++
            }
            if (heap[j]!!.frequency < heap[k]!!.frequency) {
                break
            }
            val tmp = heap[j]!!
            heap[j] = heap[k]
            heap[k]  = tmp
            j = k
        }
        return min
    }
}

fun buildTree(freqs : IntArray) : HuffTree {
    val trees = HuffHeap()
    freqs.forEachIndexed { sym, freq ->
        if (freq > 0) {
            trees.insert(HuffLeaf(freq, sym.toChar()))
        }
    }

    while (trees.size() > 1) {
        val a = trees.extract()
        val b = trees.extract()
        trees.insert(HuffNode(a, b))
    }
    return trees.extract()
}

fun buildCodes(tree: HuffTree) : Array<String> {
    val prefix = StringBuffer()
    val codes = Array<String>(maxSymbols, {""})
    buildCodes(tree, codes, prefix)
    return codes
}


fun buildCodes(tree : HuffTree, codes: Array<String>, prefix : StringBuffer) {
    when (tree) {
        is HuffLeaf -> codes[tree.symbolIndex()] = prefix.toString()

        is HuffNode -> {
            prefix.append('0')
            buildCodes(tree.left, codes, prefix)
            prefix.deleteCharAt(prefix.lastIndex)

            prefix.append('1')
            buildCodes(tree.right, codes, prefix)
            prefix.deleteCharAt(prefix.lastIndex)
        }
    }
}

File: desafio4/scala2/src/main/scala/huffman/HuffTree.scala
package huffman

import bitio.{BitInputStream, BitOutputStream}

import scala.util.control.Breaks._

/**
  * Build HuffmanTree from frequency table of symbols
  * Created by ediaz on 7/15/17.
  */
abstract class HuffTree(val frequency: Int) {

  def dumpCodes(codes : Array[String], prefix: StringBuffer)
  def writeTo(output: BitOutputStream)
  def readChar(input: BitInputStream) : Char

}

case class HuffLeaf(override val frequency: Int,  symbol: Char) extends HuffTree(frequency = frequency) {


  override def dumpCodes(codes: Array[String], prefix: StringBuffer): Unit = {
    codes(symbol & 0xFF) = prefix.toString
  }

  override def writeTo(output: BitOutputStream): Unit = {
    output.write(true)
    output.write(symbol)
  }

  override def readChar(input: BitInputStream): Char = {
    symbol
  }

}

case class HuffNode(left:HuffTree, right: HuffTree) extends HuffTree(left.frequency+right.frequency) {

  override def dumpCodes(codes: Array[String], prefix: StringBuffer): Unit = {
    prefix.append('0')
    left.dumpCodes(codes, prefix)
    prefix.deleteCharAt(prefix.length() - 1)

    prefix.append('1')
    right.dumpCodes(codes, prefix)
    prefix.deleteCharAt(prefix.length() - 1)
  }

  override def writeTo(output: BitOutputStream): Unit = {
    output.write(false)
    left.writeTo(output)
    right.writeTo(output)
  }

  override def readChar(input: BitInputStream): Char = {
    val bit = input.readBoolean()
    if (bit) right.readChar(input) else left.readChar(input)
  }
}

object HuffTree {

  val maxSymbols = 256


  def build(freqs: List[(Char, Int)]): HuffTree = {

    val leaves = freqs
      .map { case (sym, freq) => HuffLeaf(freq, sym) }.distinct
      .sortWith((l1: HuffLeaf, l2: HuffLeaf) => l1.frequency < l2.frequency)
    until(singleton, combine)(leaves).head
  }

  def singleton(trees: List[HuffTree]) : Boolean = trees.length == 1

  def combine(trees: List[HuffTree]) : List[HuffTree] =
    trees match {
      case Nil => Nil
      case left :: right :: tail => (HuffNode(left, right) :: tail).sortWith((l1,l2) => l1.frequency < l2.frequency)
    }

  def until[A](singleton: A => Boolean, combine: A => A)(data: A) : A =
    if (singleton(data)) data else until(singleton, combine)(combine(data))

  def read(input: BitInputStream): HuffTree =
    if (input.readBoolean())
      HuffLeaf(-1, input.readChar())
    else
      HuffNode(read(input), read(input))


  def buildCodes(tree: HuffTree): Array[String] = {
    val prefix = new StringBuffer()
    val codes = Array.fill[String](maxSymbols)("")
    tree.dumpCodes(codes, prefix)
    codes
  }

}